test: No direct use of nose

s-t-e-v-e-n-k opened the pull request:
## Checklist
- Tracker (select at least one)
  - [X] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [X] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [X] No doc update is appropriate
- Tests (select at least one)
  - [X] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

With nose being firmly deprecated for quite some time, remove use of it from the test suite directly.

Fixes: https://tracker.ceph.com/issues/54252

tserong commented:
jenkins test api

tserong commented:
(I'm not 100% sure who's best to review this, so, @tchaikov, @cbodley, if you're not those people, are you able to recommend someone else?)

tchaikov commented on src/test/pybind/test_ceph_argparse.py:
```diff
@@ -1330,11 +1319,16 @@ def arg_kwarg_test(self, prefix, args, sig, arg_type=0):
 #  CEPH_BIN=build/bin \
-#  PYTHONPATH=src/pybind nosetests --stop \
-#  src/test/pybind/test_ceph_argparse.py:TestOSD.test_rm"
+#  PYTHONPATH=src/pybin python3 \
```
s/pybin/pybind/

tchaikov reviewed (commented):
LGTM in general, will take a closer look tonight.

cbodley commented:
> Our primary goal is to eliminate nose from our distribution, so it is all we did.

if your only goal is to drop the package dependency, then you probably don't need to touch these rgw tests at all - they don't get packaged anywhere

alfredodeza commented:
My recommendation here would be to prioritize the work to remove nose in Ceph as a whole and work closely with the places there it is still needed to port over to Pytest. This is no longer a matter of preference, but Nose is not viable anymore as an unmaintained project. I anticipated that distributions dropping is going to be one of many problems to come if the Ceph teams decides to keep it around.

s-t-e-v-e-n-k commented:
You're welcome @cbodley! I've rebased on current master, and corrected the typo that @tchaikov spotted. I have not successfully run the `rgw_multi` tests, I don't have a suitable test environment to hand, so some help there would be lovely, and I'd be delighted to help look at some test failures caused by my changes. I've also edited `qa/tasks/rgw_multisite_tests.py` to drop nose, but I'm not certain if my changes are correct there.

cbodley commented on qa/tasks/rgw_multisite_tests.py:
```diff
@@ -79,40 +78,16 @@ def setup(self):
-
-    return LogStream()
+        argv = [sys.executable, self.module_path + "../test_multi.py"] + extra_args
+        testsuite = subprocess.run(argv, check=True)
```
i don't think this can work. `test_multi.py` is creating its own ceph clusters with the vstart.sh script, but this teuthology script is meant to run tests against the clusters that teuthology itself set up

in `qa/tasks/rgw_multisite.py`, which runs just before this `rgw_multisite_tests.py` task, we do a lot of work to set up the python environment according to teuthology's deployment, and we pass this information into the python module [here](https://github.com/ceph/ceph/blob/master/qa/tasks/rgw_multisite_tests.py#L78-L79)

so we need to run the test cases in that same environment - i don't think a subprocess can work here

cbodley commented on src/test/rgw/test_multi.py:
```diff
@@ -23,12 +22,6 @@
-from rgw_multi.tests import *
-from rgw_multi.tests_es import *
-from rgw_multi.tests_ps import *
-from rgw_multi.tests_az import *
```
i tried running this modified `test_multi.py` by hand (according to docs in https://github.com/ceph/ceph/blob/master/src/test/rgw/test_multi.md), and it isn't finding any of the test cases to run. i believe it's because the tests are no longer imported into this module
```
~/ceph/build $ cat test_multi.conf 
[DEFAULT]
num_zones = 2
num_ps_zones = 0
num_zonegroups = 1
gateways_per_zone = 1

~/ceph/build $ RGW_MULTI_TEST_CONF=test_multi.conf MON=1 OSD=1 MGR=0 MDS=0 nosetests ../src/test/rgw/test_multi.py

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
```

cbodley commented:
since these rgw tests are proving to be difficult, we might consider splitting that piece out to another PR? these other changes outside of rgw look a lot easier to verify and merge in the meantime

s-t-e-v-e-n-k commented:
What a great idea! I've split out the separate rgw test changes, and I'll create a new PR with that commit directly.

s-t-e-v-e-n-k commented:
rgw has been split out to https://github.com/ceph/ceph/pull/45759

cbodley commented:
> @cbodley ping? I thought after [#45064 (comment)](https://github.com/ceph/ceph/pull/45064#issuecomment-1083444422) and @s-t-e-v-e-n-k 's splitting the this PR, it could get merged. We really need to get rid of nose from our distro, it breaks some of our tools.

sorry, i'm only familiar with the rgw parts so assumed someone else would follow up on review here

but given @tchaikov's "LGTM in general", i'd be willing to approve this so long as it doesn't break anything in CI. i pushed this branch to ceph-ci to test the packaging at https://shaman.ceph.com/builds/ceph/pr-45064/

cbodley commented:
jenkins test make check

mcepl commented:
Thank you!
